stdlib/Makefile

```diff
@@ -22,9 +22,15 @@ OCAMLDEP ?= $(BOOT_OCAMLDEP)
 include $(ROOTDIR)/Makefile.common
 
 TARGET_BINDIR ?= $(BINDIR)
+CAMLC_FLAGS ?=
+USE_BOOT ?=
 
 COMPILER=$(ROOTDIR)/ocamlc$(EXE)
+ifeq "$(USE_BOOT)" ""
 CAMLC=$(OCAMLRUN) $(COMPILER)
+else
+CAMLC=$(BOOT_OCAMLC) $(CAMLC_FLAGS)
```

Comment:
This change looks a bit suspicious to me because the name CAMLC_FLAGS suggests that all CAMLC flags are going to go there, but most flag-passing does not use this variable at all.

It is also not clear which variable involved was previously undefined. Can you explain?

(In the previous version I am not sure how overriding `CAMLC` from the caller works if it is defined unconditionally here on line previously-27. Argh, I shouldn't look at Makefiles...)

Reply:
If you compile this `Dockerfile`:
```Dockerfile
FROM ocaml/opam:alpine-3.19-opam
RUN git clone https://github.com/ocaml/ocaml.git && git -C ocaml checkout 3908dd368eaf391f671db33804a452ad1868fb3a
WORKDIR ocaml
RUN ./configure && make -j coldstart
```
The first line of `coldstart` results in this command:
```console
make -C stdlib OCAMLRUN='$(ROOTDIR)/boot/ocamlrun' CAMLC='$(BOOT_OCAMLC) -use-prims ../runtime/primitives' all
```
Note that `$(ROOTDIR)` and `$(BOOT_OCAMLC)` are being passed _verbatim_ - i.e. they get expanded _in_ `stdlib/Makefile`, not in the root `Makefile`. If you then add `--warn-undefined-variables` to that call and run it manually, you get:
```console
$ make --warn-undefined-variables -C stdlib OCAMLRUN='$(ROOTDIR)/boot/ocamlrun' CAMLC='$(BOOT_OCAMLC) -use-prims ../runtime/primitives' all
make: Entering directory '/home/opam/ocaml/stdlib'
../Makefile.common:119: warning: undefined variable 'BOOT_OCAMLC'
make: Nothing to be done for 'all'.
make: Leaving directory '/home/opam/ocaml/stdlib'
```

Reply:
The trick I'm doing here is instead of passing the entire `CAMLC` macro and overriding the runtime (which is what the `OCAMLRUN` and `CAMLC` definitions before did), I'm passing enough information to `stdlib/Makefile` to tell it to do it. It needs to know that it's being called by coldstart and what the _additional_ flags to pass to the compiler should be. Which arguably suggests my name `CAMLC_FLAGS` is bad (open to suggestions!).

Reply:
With apologies for being slow, I still don't understand why BOOT_OCAMLC is missing there. It is defined in Makefile.common, which stdlib/Makefile correctly includes. I went to look at line 119 of Makefile.common, it reads like

```
ifeq "$(TEST_BOOT_OCAMLC_OPT)" "0"
```

and `TEST_BOOT_OCAMLC_OPT` does not seem to contain any call to CAMLC. Assuming there is an off-by-one disagreement in line numbers between make and editors, the line below is 

```
  BOOT_OCAMLC = $(ROOTDIR)/boot/ocamlc.opt
```

which also does not look like it could result in such a warning.

Reply:
Sorry, I've been unclear - I was trying to explain how the fix works, not what's presently wrong. I'm reasonably sure what's presently going wrong is a bug in GNU make 4.4.1, at the very least in the reporting of the location. Having diagnosed it as being that specific call, I elected to workaround it, as it blocks being able to run the undefined variable test in CI. I determined that if you removed the `$(BOOT_OCAMLC)` in `make --warn-undefined-variables -C stdlib OCAMLRUN='$(ROOTDIR)/boot/ocamlrun' CAMLC='$(BOOT_OCAMLC) -use-prims ../runtime/primitives' all` then the warning didn't trigger, so I therefore constructed a workaround which allowed me not to specify `$(BOOT_OCAMLC)` in that command-line.

I just tested, and the master branch of GNU make is still doing the same thing. Given that master is also reporting a fresh warning to do with our `V_` macros it may be time for a bug report, but I was trying to focus on our Windows CI, and this is blocking it 🙂